File: Algorithm/排序/希尔排序/shell_sort.py
# -*- coding:utf-8 -*-
def shell_sort(list):
	length=len(list)
	dist=int(length/2)
	while dist>0:
		for i in range(dist,length):
			temp=list[i]
			j=i
			while j>=dist and temp<list[j-dist]:
				list[j]=list[j-dist]
				j-=dist
			list[j]=temp
		dist = int(dist/2)
	return list

#测试
print("希尔排序测试")
print("排序前: ")
testbuf= [100, 7, 9, -2, 20, -50]
print(testbuf)
sorted_list = shell_sort(testbuf)
print("排序后: ")
print(sorted_list)

File: Algorithm/排序/希尔排序/shell_sort.c
#include <stdio.h>

void shell_sort(int *buf, int len)
{
   int temp = 0;
   int incre = len;

    while(1)
    {
        incre = incre/2;

        for(int k = 0; k < incre; k++)
        {
            for(int i = k+incre; i < len; i+=incre)
            {
                for(int j = i; j > k; j -= incre)
                {
                    if(buf[j] < buf[j-incre])
                    {
                        temp = buf[j - incre];
                        buf[j-incre] = buf[j];
                        buf[j] = temp;
                    }
                    else
                    {
                        break;
                    }
                }
            }
        }

        if(incre == 1)
        {
            break;
        }
    }
}

int testbuf[] = {100, 7, 9, -2, 20, -50};

int main(void)
{
    int testlen = sizeof(testbuf)/sizeof(int);
    printf("选择排序测试\n");
    printf("排序前: \n");
    for(int i = 0; i < testlen; i++)
    {
        printf("%d ", testbuf[i]);
    }
    printf("\n");
    shell_sort(testbuf, testlen);
    printf("排序后: \n");
    for(int i = 0; i < testlen; i++)
    {
        printf("%d ", testbuf[i]);
    }
    printf("\n");
}